modules/services/hercules-ci-agent/default.nix

```diff
@@ -0,0 +1,104 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.hercules-ci-agent;
+  user = config.users.users.hercules-ci-agent;
+in {
+  options.services.hercules-ci-agent = {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = "Run Hercules CI Agent as a system service";
+    };
+
+    apiBaseUrl = mkOption {
+      description = "Alternative base URL for the Hercules API";
+      example = "https://hercules-ci.com";
+      default = null;
+      type = types.nullOr types.string;
+    };
+
+    clusterJoinTokenPath = mkOption {
+      description = ''
+        Location of a the cluster join token. It authorizes the agent to add
+        itself to the cluster that the token represents.
+
+        This file is only required to be present for the agent's first run. It
+        will be ignored after the agent has used the token successfully.
+      '';
+      type = types.path;
+    };
+
+    logFile = mkOption {
+      type = types.path;
+      default = "/var/log/hercules-ci-agent.log";
+      description = "Stdout and sterr of hercules-ci-agent process.";
+    };
+
+    concurrentTasks = mkOption {
+      description = "Number of tasks to perform simultaneously, such as evaluations, derivations";
+      type = types.int;
+      default = 4;
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = (import (builtins.fetchTarball "https://github.com/hercules-ci/hercules-ci-agent/archive/stable.tar.gz")).nixos-19_03.hercules-ci-agent;
```

Comment:
I'd very much prefer to have a default package with IFD than instruct users to copy/paste :)

Let me know if that's something you find acceptable, otherwise, I'll include it into the docs.

Reply:
I have 2 main problems with this.
1. It makes the system not reproducable, which is fine if that makes the most sense for the service, but shouldn't happen implicitly.
2. Fetching 3rd party expressions has security implications, something like this would also have to be explicit for nixos IMHO.

Reply:
I share those two problems as well. I'll switch to providing an example instead.

However, I'd like to point out that making users copy paste a line manually doesn't solve much and the problems remain.

Third party expressions are a matter of trust, which means the circle is bigger than just trusting your expressions. I honestly think that's the case anyway for those that will want to use hercules-ci, they will have to trust us (which has same guarantees as nix-darwin, being open source code).

Reply:
Yeah it doesn't solve anything except that it's not hidden inside the module. Perhaps the channel 2.0 work that's going on will provide a better alternative for fetching 3rd party sources.